force_gromacs/data_sources/fragment/fragment_data_source.py

```diff
@@ -10,20 +14,38 @@ class FragmentDataSource(BaseDataSource):
     are not necessary for atoms or molecules represented by single beads.
     """
 
+    _reader = GromacsMoleculeReader()
+
     def run(self, model, parameters):
         """Simply wraps all user input in a `GromacsFragment` object for further
         processing. Consequently, it is expected that either this method
         can be overloaded by a subclass to perform more specific actions,
         of additional `DataSource` objects can perform this in the next
         `ExecutionLayer`"""
 
-        fragment = GromacsFragment(
-            name=model.name,
-            symbol=model.symbol,
-            topology=model.topology,
-            coordinate=model.coordinate
+        # Obtain all fragments present in topology file
+        fragments = self._reader.read(
+            model.topology
         )
 
+        # Search for fragment with symbol referenced in model
+        indices = [
+            index for index, fragment in enumerate(fragments)
+            if fragment.symbol == model.symbol
+        ]
+        if indices:
+            fragment = fragments[indices[0]]
+        else:
```

Comment:
Would it be more transparent to use `try / except` instead of `if / else`?

Reply:
Yes, agreed - can raise a new exception from `IndexError`